crates/sui-indexer-alt/src/queries/objects.rs

```diff
@@ -0,0 +1,246 @@
+    /// name.
+    ///
+    /// Generic types can be queried by either the generic type name, e.g. `0x2::coin::Coin`, or by
+    /// the full type name, such as `0x2::coin::Coin<0x2::sui::SUI>`.
+    pub type_filter: Option<TypeFilter>,
+
+    /// Filter for live objects by their current owners.
+    pub owner_filter: Option<SuiAddress>,
+}
+
+#[derive(QueryableByName, Selectable, Debug)]
+#[diesel(table_name = obj_info)]
+struct IdCheckpoint {
+    #[diesel(sql_type = BigInt)]
+    cp_sequence_number: i64,
+    #[diesel(sql_type = Bytea)]
+    object_id: Vec<u8>,
+}
+
+pub async fn query_objects_with_filters(
+    db: &Db,
+    view_checkpoint_number: i64,
+    filters: ObjectFilter,
+    cursor: Option<Cursor>,
+    limit: usize,
+) -> anyhow::Result<Vec<StoredObjVersion>> {
+    let object_ids =
+        query_object_ids_with_filters(db, view_checkpoint_number, filters, cursor, limit).await?;
+    query_latest_object_versions(db, &object_ids).await
+}
+
+// TODO: Double check that this function is not prone to SQL injection.
+async fn query_object_ids_with_filters(
+    db: &Db,
+    view_checkpoint_number: i64,
+    filters: ObjectFilter,
+    cursor: Option<Cursor>,
+    limit: usize,
+) -> anyhow::Result<Vec<IdCheckpoint>> {
+    let mut filter_conditions = vec![];
+    if let Some(owner) = filters.owner_filter {
+        filter_conditions.push(format!("owner_kind = {}", StoredOwnerKind::Address as i16));
+        filter_conditions.push(format!(
+            "owner_id = '\\x{}'::bytea",
+            hex::encode(owner.to_vec())
+        ));
+    }
+
+    if let Some(type_filter) = filters.type_filter {
+        match type_filter {
+            TypeFilter::ByModule(module) => match module {
+                ModuleFilter::ByPackage(package) => {
+                    filter_conditions.push(format!(
+                        "package = '\\x{}'::bytea",
+                        hex::encode(package.to_vec())
+                    ));
+                }
+                ModuleFilter::ByModule(package, module) => {
+                    filter_conditions.push(format!(
+                        "package = '\\x{}'::bytea",
+                        hex::encode(package.to_vec()),
+                    ));
+                    filter_conditions.push(format!("module = '{}'", module));
+                }
+            },
+            TypeFilter::ByType(struct_tag) => {
+                filter_conditions.push(format!(
+                    "package = '\\x{}'::bytea",
+                    hex::encode(struct_tag.address.to_vec())
+                ));
+                filter_conditions.push(format!("module = '{:?}'", struct_tag.module.as_str()));
+                filter_conditions.push(format!("name = '{:?}'", struct_tag.name.as_str()));
+                filter_conditions.push(format!(
+                    "instantiation = '\\x{}'::bytea",
+                    hex::encode(bcs::to_bytes(&struct_tag.type_params).unwrap())
```

Comment:
When `struct_tag.type_params` is empty, this is going to be encoded into empty bytes possibly. I'm not sure if that'll be equal to NULL in postgres.

Reply:
That's intended though. When we store the type_params, for empty type_params, we store the bcs encoding of the empty vector, not NULL.